lib/versioning/VersioningRequestProcessor.js

```diff
@@ -0,0 +1,404 @@
+            return callback(null, [{ key, value }], versionId);
+        }
+        // need to get the master version to check if this is the master version
+        this.writeCache.get({ db, key }, logger, (err, data) => {
+            if (err && !err.ObjNotFound) {
+                return callback(err);
+            }
+            const versionId = request.options.versionId;
+            const versionKey = formatVersionKey(request.key, versionId);
+            const ops = [{ key: versionKey, value: request.value }];
+            if (Version.from(data).getVersionId() === versionId) {
+                // the version is also the master version: need to update both
+                ops.push({ key: request.key, value: request.value });
+            }
+            return callback(null, ops, versionId);
+        });
+        return null;
+    }
+
+
+    del(request, logger, callback) {
+        const { db, key, options } = request;
+        // no versioning or versioning configuration off
+        if (!(options && options.versionId)) {
+            return this.writeCache.batch({ db,
+                                           array: [{ key, type: 'del' }] },
+                                         logger, callback);
+        }
+        // version specific DELETE
+        return this.processVersionSpecificDelete(request, logger,
+            (err, array, vid) => {
+                if (err) {
+                    return callback(err);
+                }
+                return this.writeCache.batch({ db, array }, logger,
+                    err => callback(err, `{"versionId":"${vid}"}`));
+            });
+    }
+
+    /**
+     * Processes a version specific deleteObject request. This will create a
+     * batch of operations for deleting the specific version and marking the
+     * master version of the object as a place holder for deletion if the
+     * specific version is also the master version.
+     *
+     * @param {object} request - the request in original
+     *                           RepdConnection format { db, key
+     *                           [, value][, type], method, options }
+     * @param {object} logger - logger
+     * @param {function} callback - expect callback(err, batch, versionId)
+     * @return {any} - to finish the call
+     */
+    processVersionSpecificDelete(request, logger, callback) {
+        const { db, key, options } = request;
+        // deleting a specific version
+        this.writeCache.get({ db, key }, logger, (err, data) => {
+            if (err && !err.ObjNotFound) {
+                return callback(err);
+            }
+            // delete the specific version
+            const versionId = options.versionId;
+            const versionKey = formatVersionKey(key, versionId);
+            const ops = [{ key: versionKey, type: 'del' }];
+            // update the master version as PHD if it is the deleting version
+            if (Version.isPHD(data) ||
```

Comment:
Note that I previously used a non-versioning-aware branch when doing a batch of deletes, so it may have caused inconsistency in my local store. The ObjNotFound error should be handled gracefully though.